include/lsst/afw/table/pybind11Catalog.h

```diff
@@ -0,0 +1,194 @@
+        if (iter == self.end()) {
+            return PTR(RecordT)();
+        };
+        return iter;
+    });
+    cls.def(("_upper_bound_"+suffix).c_str(), [](Catalog & self,
+                                                  T const & value, Key<T> const & key)->int{
+        return self.upper_bound(value, key) - self.begin();
+    });
+    cls.def(("_lower_bound_"+suffix).c_str(), [](Catalog & self,
+                                                  T const & value, Key<T> const & key)->int{
+        return self.lower_bound(value, key) - self.begin();
+    });
+    cls.def(("_equal_range_"+suffix).c_str(), [](Catalog & self,
+                                                  T const & value, Key<T> const & key)->std::pair<int,int>{
+        std::pair<typename Catalog::const_iterator,typename Catalog::const_iterator> p
+            = self.equal_range(value, key);
+        return std::pair<int,int>(p.first - self.begin(), p.second - self.begin());
+    });
+};
+
+/**
+Declare member and static functions for a given instantiation of lsst::afw::table::CatalogT<RecordT>.
+
+@tparam RecordT  Record type, e.g. BaseRecord or SimpleRecord.
+
+@param[in] cls  Catalog pybind11 class.
+*/
+template <typename RecordT>
+void declareCatalog(py::class_<CatalogT<RecordT>, std::shared_ptr<CatalogT<RecordT>>> & cls) {
+    typedef CatalogT<RecordT> Catalog;
+    typedef typename RecordT::Table Table;
+    typedef typename RecordT::ColumnView ColumnView;
+    typedef std::vector<PTR(RecordT)> Internal;
+    typedef typename Internal::size_type size_type;
+    typedef RecordT & reference;
+
+    /* Constructors */
+    cls.def(py::init<Schema const &>());
+    cls.def(py::init<PTR(Table) const &>());
+    cls.def(py::init<Catalog const &>());
+
+    /* Members */
+    cls.def("getTable", &Catalog::getTable);
+    cls.def("getSchema", &Catalog::getSchema);
+    cls.def("getColumnView", &Catalog::getColumnView);
+    cls.def("capacity", &Catalog::capacity);
+    cls.def("addNew", &Catalog::addNew);
+    cls.def("__len__", &Catalog::size);
+    cls.def("set", &Catalog::set);
+    cls.def("_getitem_", [](Catalog & self, int i) {
+        // If the index is less than 0, use the pythonic index
+        if(i<0){
+            i = self.size()+i;
+        };
+        return self.get(i);
+    });
+    cls.def("isContiguous", &Catalog::isContiguous);
+    cls.def("writeFits",
+                    (void (Catalog::*)(std::string const &, std::string const &, int) const)
+                        &Catalog::writeFits,
+                    "filename"_a, "mode"_a="w", "flags"_a = 0);
+    cls.def("writeFits",
+                    (void (Catalog::*)(fits::MemFileManager &,
+                                                 std::string const &,
+                                                 int) const) &Catalog::writeFits,
+                    "manager"_a, "mode"_a="w", "flags"_a=0);
+    //cls.def("writeFits", (void (Catalog::*)() const) &Catalog::writeFits);
```

Comment:
I have been intentionally leaving other methods wrapped but commented out so they can trivially be enabled if/when the need for them appears. This had been done in some of the existing pybind11 code and I found it *very* useful -- it made my job much faster. I hope we can continue to do this. I agree we should have a final pass at some point to remove all the remaining lines or add tests for them and uncomment them.